packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -966,35 +1006,89 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
 
   final Map<int, Widget> _childWidgets = HashMap<int, Widget>();
   final SplayTreeMap<int, Element> _childElements = SplayTreeMap<int, Element>();
-  RenderBox _currentBeforeChild;
+
 
   @override
   void performRebuild() {
+    renderObject.integrityEnabled = false; // moving children will temporary violate the integrity
     _childWidgets.clear(); // Reset the cache, as described above.
     super.performRebuild();
-    _currentBeforeChild = null;
     assert(_currentlyUpdatingChildIndex == null);
     try {
+      final Set<int> affectedIndexes = {};
+      final Map<int, Element> newKeepAliveChildren = <int, Element>{};
+
+      int getNewIndex(Widget w){
+        if (w.key != null){
+          return widget.delegate.getIndexByKey(w.key);
+        } else {
+          return -1;
+        }
+      }
+
       void processElement(int index) {
         _currentlyUpdatingChildIndex = index;
         final Element newChild = updateChild(_childElements[index], _build(index), index);
         if (newChild != null) {
           _childElements[index] = newChild;
-          final SliverMultiBoxAdaptorParentData parentData = newChild.renderObject.parentData;
-          if (!parentData.keptAlive)
-            _currentBeforeChild = newChild.renderObject;
         } else {
           _childElements.remove(index);
         }
       }
-      // processElement may modify the Map - need to do a .toList() here.
-      _childElements.keys.toList().forEach(processElement);
+
+      void removeChildIfNotRemapped(int index){
+        final Element child = _childElements[index];
+        final SliverMultiBoxAdaptorParentData originalParentData = child?.renderObject?.parentData;
+        if (originalParentData != null){
+          if (!originalParentData.keepAlive || getNewIndex(
+              child.widget) == -1) {
+            _childElements[index] = updateChild(_childElements[index], null, index);
+          }
+        }
+      }
+
+      void remappingKeepAliveChildren(int index) {
+        _currentlyUpdatingChildIndex = index;
+        if (_childElements[index] != null) {
+          removeChildIfNotRemapped(index);
+        }
+        final Element newChild = updateChild(newKeepAliveChildren[index], _build(index), index);
+        if (newChild != null) {
+          _childElements[index] = newChild;
+        } else {
+          _childElements.remove(index);
+        }
+      }
+
+      void remappingKeepAliveChildIndexes(int index){
```

Comment:
I think we want to remap all children that have a Key (not just those that were kept alive). 

The general contract for a keyed widget is that if the widget moves around in the child list, we will also move the associated element/renderobject to the new place. Sliver are currently in violation of that and this would fix it.

Reply:
refactored